File: 프로그래머스/Level 2/연속된 부분 수열의 합.cpp
// https://school.programmers.co.kr/learn/courses/30/lessons/178870
#include <iostream>
#include <string>
#include <vector>

using namespace std;

vector<int> solution(vector<int> sequence, int k) {
    vector<int> answer;
    int mindiff = 1e9;
    int seq_total = sequence[0], left = 0, right = 0;
    
    while (left < sequence.size() && right < sequence.size()) 
    {
        if (seq_total == k && mindiff > right-left)
        {
            mindiff = right-left;
            answer = {left, right};
        }
        //cout << left << ' ' << right << '\n';
        if (seq_total > k)
        {
            seq_total -= sequence[left++];
        } else
        {
            ++right;
            if (right >= sequence.size()) continue;
            seq_total += sequence[right];
        }
    }
    return answer;
}

File: 프로그래머스/Level 2/연속된 부분 수열의 합.swift
// https://school.programmers.co.kr/learn/courses/30/lessons/178870
import Foundation

func solution(_ sequence:[Int], _ k:Int) -> [Int] {
    var answer: [Int] = []
    var l = 0
    var r = 0
    var tot = sequence[r]
    var maxDiff = 1000001
    
    while (l < sequence.count && r < sequence.count) {        
        if tot > k { // 왼쪽
            tot -= sequence[l]
            if l + 1 >= sequence.count { break }
            l += 1
            
        } else { // 오른쪽
            if tot == k &&  maxDiff > r - l {
                answer = [l, r]
                maxDiff = r - l
            }
            
            if r + 1 >= sequence.count { break } // 범위 확인
            r += 1
            tot += sequence[r] // illegal instruction (core dumped)?
        }
    }
    return answer
}